File: 백준/Gold/1300. K번째 수/K번째 수.java
import java.io.*;

public class Main {

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int N = Integer.parseInt(br.readLine());
        int K = Integer.parseInt(br.readLine());

        long left = 1;
        long right = K;

        while (left < right) {
            long mid = (left + right) / 2;
            long cnt = 0;

            for (int i = 1; i <= N; i++) {
                cnt += Math.min(mid / i, N);
            }

            if (K <= cnt) {
                right = mid;
            } else {
                left = mid + 1;
            }
        }

        System.out.println(right);
    }
}

File: 백준/Gold/1300. K번째 수/K번째 수.py
import sys

def bn_search(l, r):
    ans = 0
    while l <= r:
        mid = (l + r) // 2
        sum = 0
        
        for i in range(1, N + 1):
            sum += min(mid // i, N)
        
        if sum >= k:
            ans = mid
            r = mid - 1
        else:
            l = mid + 1
    print(ans)

N = int(sys.stdin.readline())

k = int(sys.stdin.readline())

bn_search(1, N*N)